File: resqradar/flutter_app/lib/dsp/signal_filters.dart
import 'dart:math' as math;

class FftUtils {
  static int nextPowerOfTwo(int n) {
    int v = 1;
    while (v < n) v = v << 1;
    return v;
  }

  static Map<String, List<double>> magnitudeSpectrum(List<double> signal, double fs) {
    int n = nextPowerOfTwo(signal.length);
    List<double> re = List.filled(n, 0.0);
    List<double> im = List.filled(n, 0.0);
    for (int i = 0; i < signal.length; i++) {
      re[i] = signal[i];
    }
    _fftInPlace(re, im);

    int half = n ~/ 2;
    List<double> freqs = List.filled(half, 0.0);
    List<double> mags = List.filled(half, 0.0);
    for (int k = 0; k < half; k++) {
      freqs[k] = k * fs / n;
      mags[k] = math.sqrt(re[k] * re[k] + im[k] * im[k]);
    }
    return {'freqs': freqs, 'mags': mags};
  }

  static void _fftInPlace(List<double> re, List<double> im) {
    int n = re.length;
    int j = 0;
    for (int i = 1; i < n; i++) {
      int bit = n >> 1;
      while ((j & bit) != 0) {
        j ^= bit;
        bit >>= 1;
      }
      j ^= bit;
      if (i < j) {
        double tempRe = re[i];
        re[i] = re[j];
        re[j] = tempRe;
        double tempIm = im[i];
        im[i] = im[j];
        im[j] = tempIm;
      }
    }

    int len = 2;
    while (len <= n) {
      double ang = -2.0 * math.pi / len;
      double wLenRe = math.cos(ang);
      double wLenIm = math.sin(ang);
      int i = 0;
      while (i < n) {
        double wRe = 1.0;
        double wIm = 0.0;
        for (int k = 0; k < len ~/ 2; k++) {
          double uRe = re[i + k];
          double uIm = im[i + k];
          double vRe = re[i + k + len ~/ 2] * wRe - im[i + k + len ~/ 2] * wIm;
          double vIm = re[i + k + len ~/ 2] * wIm + im[i + k + len ~/ 2] * wRe;
          re[i + k] = uRe + vRe;
          im[i + k] = uIm + vIm;
          re[i + k + len ~/ 2] = uRe - vRe;
          im[i + k + len ~/ 2] = uIm - vIm;
          double nextWRe = wRe * wLenRe - wIm * wLenIm;
          wIm = wRe * wLenIm + wIm * wLenRe;
          wRe = nextWRe;
        }
        i += len;
      }
      len <<= 1;
    }
  }

  static SpectralMetrics spectralMetrics(List<double> signal, double fs, List<double> band) {
    var spectrum = magnitudeSpectrum(signal, fs);
    List<double> freqs = spectrum['freqs']!;
    List<double> mags = spectrum['mags']!;
    
    double bestF = 0.0;
    double bestM = 0.0;
    double inBand = 0.0;
    double outBand = 0.0;
    double total = 0.0;
    
    for (int i = 0; i < freqs.length; i++) {
      double e = mags[i] * mags[i];
      total += e;
      if (freqs[i] >= band[0] && freqs[i] <= band[1]) {
        inBand += e;
        if (mags[i] > bestM) {
          bestM = mags[i];
          bestF = freqs[i];
        }
      } else {
        outBand += e;
      }
    }
    double specEnergy = (total < 1e-12) ? 0.0 : inBand / total;
    double snr = (outBand < 1e-12) ? 100.0 : inBand / outBand;
    return SpectralMetrics(bestF, bestM, specEnergy, snr);
  }
}

class SpectralMetrics {
  final double dominantFreq;
  final double dominantMag;
  final double spectralEnergy;
  final double snr;
  SpectralMetrics(this.dominantFreq, this.dominantMag, this.spectralEnergy, this.snr);
}

class SignalFilters {
  static List<double> hampelFilter(List<double> signal, {int windowSize = 5, double threshold = 3.5}) {
    List<double> out = List.from(signal);
    double k = 1.4826;
    for (int i = 0; i < signal.length; i++) {
      int lo = math.max(0, i - windowSize);
      int hi = math.min(signal.length, i + windowSize + 1);
      List<double> window = signal.sublist(lo, hi)..sort();
      double median = window[window.length ~/ 2];
      
      List<double> absDevs = window.map((e) => (e - median).abs()).toList()..sort();
      double mad = k * absDevs[absDevs.length ~/ 2];
      
      if (mad > 0 && (signal[i] - median).abs() > threshold * mad) {
        out[i] = median;
      }
    }
    return out;
  }

  static List<double> removeSlowTrend(List<double> signal, double fs, double windowSec) {
    int win = math.max(3, (fs * windowSec).toInt());
    if (signal.length < win) return List.from(signal);
    List<double> out = List.from(signal);
    int half = win ~/ 2;
    for (int i = 0; i < signal.length; i++) {
      int lo = math.max(0, i - half);
      int hi = math.min(signal.length, i + half + 1);
      double sum = 0;
      for (int j = lo; j < hi; j++) sum += signal[j];
      double mean = sum / (hi - lo);
      out[i] -= mean;
    }
    return out;
  }

  static List<double> bandpass(List<double> signal, double lowHz, double highHz, double fs) {
    List<double> detrended = removeSlowTrend(signal, fs, math.max(5.0, 1.0 / lowHz));
    return _lowPass(detrended, highHz, fs);
  }

  static List<double> _lowPass(List<double> signal, double cutoffHz, double fs) {
    double rc = 1.0 / (2.0 * math.pi * cutoffHz);
    double dt = 1.0 / fs;
    double alpha = dt / (rc + dt);
    List<double> out = List.filled(signal.length, 0.0);
    out[0] = signal[0];
    for (int i = 1; i < signal.length; i++) {
      out[i] = out[i - 1] + alpha * (signal[i] - out[i - 1]);
    }
    return out;
  }

  static List<double> preprocess(List<double> signal, double fs, bool rubbleMode) {
    double avg = signal.isEmpty ? 0 : signal.reduce((a, b) => a + b) / signal.length;
    List<double> x = signal.map((e) => e - avg).toList();
    
    List<double> d1 = List.filled(x.length, 0.0);
    for (int i = 0; i < x.length; i++) {
      d1[i] = (i == 0) ? x[0] : x[i] - x[i - 1];
    }
    double d1Mean = d1.reduce((a, b) => a + b) / d1.length;
    for (int i = 0; i < d1.length; i++) d1[i] -= d1Mean;

    if (!rubbleMode) {
      return List.generate(x.length, (i) => 0.65 * x[i] + 0.35 * d1[i]);
    }

    List<double> d2 = List.filled(d1.length, 0.0);
    for (int i = 0; i < d1.length; i++) {
      d2[i] = (i == 0) ? d1[0] : d1[i] - d1[i - 1];
    }
    double d2Mean = d2.reduce((a, b) => a + b) / d2.length;
    for (int i = 0; i < d2.length; i++) d2[i] -= d2Mean;

    int win = math.max(5, (fs * 15).toInt());
    List<double> trendRemoved = x;
    if (x.length >= win) {
      trendRemoved = List.filled(x.length, 0.0);
      for (int i = 0; i < x.length; i++) {
        int lo = math.max(0, i - win ~/ 2);
        int hi = math.min(x.length, i + win ~/ 2 + 1);
        double sum = 0;
        for (int j = lo; j < hi; j++) sum += x[j];
        double mean = sum / (hi - lo);
        trendRemoved[i] = x[i] - mean;
      }
    }
    
    return List.generate(trendRemoved.length, (i) => 0.45 * trendRemoved[i] + 0.35 * d1[i] + 0.20 * d2[i]);
  }

  static double respiratoryPeriodicity(List<double> filtered, double fs) {
    if (filtered.length < math.max(30, (fs * 12).toInt())) return 0.0;
    double mean = filtered.reduce((a, b) => a + b) / filtered.length;
    List<double> centered = filtered.map((e) => e - mean).toList();
    List<double> ac = _correlate(centered, centered);
    if (ac.isEmpty || ac[0] <= 1e-12) return 0.0;
    
    List<double> norm = ac.map((e) => e / ac[0]).toList();
    int lagMin = math.max(1, (fs / 0.55).toInt());
    int lagMax = math.min(norm.length - 1, (fs / 0.06).toInt());
    if (lagMax <= lagMin) return 0.0;
    
    double maxVal = -double.maxFinite;
    for (int i = lagMin; i < lagMax; i++) {
      if (norm[i] > maxVal) maxVal = norm[i];
    }
    return maxVal;
  }

  static List<double> _correlate(List<double> a, List<double> b) {
    int n = a.length;
    List<double> out = List.filled(2 * n - 1, 0.0);
    for (int i = 0; i < out.length; i++) {
      double sum = 0.0;
      for (int j = 0; j < a.length; j++) {
        int k = i - (n - 1) + j;
        if (k >= 0 && k < b.length) sum += a[j] * b[k];
      }
      out[i] = sum;
    }
    return out.sublist(n - 1);
  }

  static int uniqueRounded(List<double> signal, {int decimals = 2}) {
    double factor = math.pow(10.0, decimals.toDouble()).toDouble();
    Set<double> unique = {};
    for (double v in signal) {
      unique.add((v * factor).round() / factor);
    }
    return unique.length;
  }

  static double variance(List<double> signal) {
    if (signal.isEmpty) return 0.0;
    double mean = signal.reduce((a, b) => a + b) / signal.length;
    double sumSq = 0;
    for (double v in signal) sumSq += (v - mean) * (v - mean);
    return sumSq / signal.length;
  }
}

File: resqradar/android/app/src/main/java/com/whofi/vitalfi/dsp/SignalFilters.kt
package com.whofi.vitalfi.dsp

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.exp
import kotlin.math.floor
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.min
import kotlin.math.round
import kotlin.math.sin
import kotlin.math.sqrt

object FftUtils {

    fun nextPowerOfTwo(n: Int): Int {
        var v = 1
        while (v < n) v = v shl 1
        return v
    }

    /** FFT radix-2 Cooley-Tukey. Entrada real, salida magnitudes por bin positivo. */
    fun magnitudeSpectrum(signal: DoubleArray, fs: Double): Pair<DoubleArray, DoubleArray> {
        val n = nextPowerOfTwo(signal.size)
        val re = DoubleArray(n)
        val im = DoubleArray(n)
        for (i in signal.indices) re[i] = signal[i]
        fftInPlace(re, im)

        val half = n / 2
        val freqs = DoubleArray(half) { k -> k * fs / n }
        val mags = DoubleArray(half) { k ->
            sqrt(re[k] * re[k] + im[k] * im[k])
        }
        return freqs to mags
    }

    private fun fftInPlace(re: DoubleArray, im: DoubleArray) {
        val n = re.size
        var j = 0
        for (i in 1 until n) {
            var bit = n shr 1
            while (j and bit != 0) {
                j = j xor bit
                bit = bit shr 1
            }
            j = j xor bit
            if (i < j) {
                re[i] = re[j].also { re[j] = re[i] }
                im[i] = im[j].also { im[j] = im[i] }
            }
        }

        var len = 2
        while (len <= n) {
            val ang = -2.0 * PI / len
            val wLenRe = cos(ang)
            val wLenIm = sin(ang)
            var i = 0
            while (i < n) {
                var wRe = 1.0
                var wIm = 0.0
                for (k in 0 until len / 2) {
                    val uRe = re[i + k]
                    val uIm = im[i + k]
                    val vRe = re[i + k + len / 2] * wRe - im[i + k + len / 2] * wIm
                    val vIm = re[i + k + len / 2] * wIm + im[i + k + len / 2] * wRe
                    re[i + k] = uRe + vRe
                    im[i + k] = uIm + vIm
                    re[i + k + len / 2] = uRe - vRe
                    im[i + k + len / 2] = uIm - vIm
                    val nextWRe = wRe * wLenRe - wIm * wLenIm
                    wIm = wRe * wLenIm + wIm * wLenRe
                    wRe = nextWRe
                }
                i += len
            }
            len = len shl 1
        }
    }

    fun dominantFrequency(
        signal: DoubleArray,
        fs: Double,
        band: Pair<Double, Double>,
    ): Pair<Double, Double> {
        val metrics = spectralMetrics(signal, fs, band)
        return metrics.dominantFreq to metrics.dominantMag
    }

    data class SpectralMetrics(
        val dominantFreq: Double,
        val dominantMag: Double,
        val spectralEnergy: Double,
        val snr: Double,
    )

    fun spectralMetrics(
        signal: DoubleArray,
        fs: Double,
        band: Pair<Double, Double>,
    ): SpectralMetrics {
        val (freqs, mags) = magnitudeSpectrum(signal, fs)
        var bestF = 0.0
        var bestM = 0.0
        var inBand = 0.0
        var outBand = 0.0
        var total = 0.0
        for (i in freqs.indices) {
            val e = mags[i] * mags[i]
            total += e
            if (freqs[i] in band.first..band.second) {
                inBand += e
                if (mags[i] > bestM) {
                    bestM = mags[i]
                    bestF = freqs[i]
                }
            } else {
                outBand += e
            }
        }
        val specEnergy = if (total < 1e-12) 0.0 else inBand / total
        val snr = if (outBand < 1e-12) 100.0 else inBand / outBand
        return SpectralMetrics(bestF, bestM, specEnergy, snr)
    }

    fun spectralEnergy(signal: DoubleArray, fs: Double, band: Pair<Double, Double>): Double =
        spectralMetrics(signal, fs, band).spectralEnergy

    fun signalSnr(signal: DoubleArray, fs: Double, band: Pair<Double, Double>): Double =
        spectralMetrics(signal, fs, band).snr
}

object SignalFilters {

    fun hampelFilter(signal: DoubleArray, windowSize: Int = 5, threshold: Double = 3.5): DoubleArray {
        val out = signal.copyOf()
        val k = 1.4826
        for (i in signal.indices) {
            val lo = max(0, i - windowSize)
            val hi = min(signal.size, i + windowSize + 1)
            val window = signal.slice(lo until hi).sorted()
            val median = window[window.size / 2]
            val mad = k * window.map { abs(it - median) }.sorted()[window.size / 2]
            if (mad > 0 && abs(signal[i] - median) > threshold * mad) {
                out[i] = median
            }
        }
        return out
    }

    /** Bandpass aproximado: quita tendencia lenta + suavizado paso-bajo. */
    fun bandpass(signal: DoubleArray, lowHz: Double, highHz: Double, fs: Double): DoubleArray {
        val detrended = removeSlowTrend(signal, fs, max(5.0, 1.0 / lowHz))
        return lowPass(detrended, cutoffHz = highHz, fs = fs)
    }

    fun removeSlowTrend(signal: DoubleArray, fs: Double, windowSec: Double): DoubleArray {
        val win = max(3, (fs * windowSec).toInt())
        if (signal.size < win) return signal.copyOf()
        val out = signal.copyOf()
        val half = win / 2
        for (i in signal.indices) {
            val lo = max(0, i - half)
            val hi = min(signal.size, i + half + 1)
            val mean = signal.slice(lo until hi).average()
            out[i] -= mean
        }
        return out
    }

    private fun lowPass(signal: DoubleArray, cutoffHz: Double, fs: Double): DoubleArray {
        val rc = 1.0 / (2.0 * PI * cutoffHz)
        val dt = 1.0 / fs
        val alpha = dt / (rc + dt)
        val out = DoubleArray(signal.size)
        out[0] = signal[0]
        for (i in 1 until signal.size) {
            out[i] = out[i - 1] + alpha * (signal[i] - out[i - 1])
        }
        return out
    }

    fun preprocess(signal: DoubleArray, fs: Double, rubbleMode: Boolean): DoubleArray {
        val x = signal.map { it - signal.average() }.toDoubleArray()
        val d1 = DoubleArray(x.size) { i ->
            val v = if (i == 0) x[0] else x[i] - x[i - 1]
            v
        }
        val d1Mean = d1.average()
        for (i in d1.indices) d1[i] -= d1Mean

        if (!rubbleMode) {
            return DoubleArray(x.size) { i -> 0.65 * x[i] + 0.35 * d1[i] }
        }

        val d2 = DoubleArray(d1.size) { i ->
            val v = if (i == 0) d1[0] else d1[i] - d1[i - 1]
            v
        }
        val d2Mean = d2.average()
        for (i in d2.indices) d2[i] -= d2Mean

        val win = max(5, (fs * 15).toInt())
        var trendRemoved = x
        if (x.size >= win) {
            trendRemoved = DoubleArray(x.size)
            for (i in x.indices) {
                val lo = max(0, i - win / 2)
                val hi = min(x.size, i + win / 2 + 1)
                val mean = x.slice(lo until hi).average()
                trendRemoved[i] = x[i] - mean
            }
        }
        return DoubleArray(trendRemoved.size) { i ->
            0.45 * trendRemoved[i] + 0.35 * d1[i] + 0.20 * d2[i]
        }
    }

    fun respiratoryPeriodicity(filtered: DoubleArray, fs: Double): Double {
        if (filtered.size < max(30, (fs * 12).toInt())) return 0.0
        val mean = filtered.average()
        val centered = filtered.map { it - mean }.toDoubleArray()
        val ac = correlate(centered, centered)
        if (ac.isEmpty() || ac[0] <= 1e-12) return 0.0
        val norm = ac.map { it / ac[0] }
        val lagMin = max(1, (fs / 0.55).toInt())
        val lagMax = min(norm.size - 1, (fs / 0.06).toInt())
        if (lagMax <= lagMin) return 0.0
        return norm.subList(lagMin, lagMax).maxOrNull() ?: 0.0
    }

    private fun correlate(a: DoubleArray, b: DoubleArray): DoubleArray {
        val n = a.size
        val out = DoubleArray(2 * n - 1)
        for (i in out.indices) {
            var sum = 0.0
            for (j in a.indices) {
                val k = i - (n - 1) + j
                if (k in b.indices) sum += a[j] * b[k]
            }
            out[i] = sum
        }
        return out.slice(n - 1 until out.size).toDoubleArray()
    }

    fun uniqueRounded(signal: DoubleArray, decimals: Int = 2): Int {
        val factor = Math.pow(10.0, decimals.toDouble())
        return signal.map { kotlin.math.round(it * factor) / factor }.toSet().size
    }

    fun variance(signal: DoubleArray): Double {
        if (signal.isEmpty()) return 0.0
        val mean = signal.average()
        return signal.map { (it - mean) * (it - mean) }.average()
    }
}
